src/Dfc.CourseDirectory.WebV2/Features/Venues/DeleteVenue/DeleteVenue.cs

```diff
@@ -0,0 +1,232 @@
+        public Guid CourseRunId { get; set; }
+        public string CourseName { get; set; }
+    }
+
+    public class AffectedApprenticeshipViewModel
+    {
+        public Guid ApprenticeshipId { get; set; }
+        public string ApprenticeshipName { get; set; }
+    }
+
+    public class AffectedTLevelViewModel
+    {
+        public Guid TLevelId { get; set; }
+        public string TLevelName { get; set; }
+    }
+
+    public class DeletedQuery : IRequest<DeletedViewModel>
+    {
+    }
+
+    public class DeletedViewModel
+    {
+        public string VenueName { get; set; }
+    }
+
+    public class Handler :
+        IRequestHandler<Query, OneOf<NotFound, ViewModel>>,
+        IRequestHandler<Command, OneOf<NotFound, ModelWithErrors<ViewModel>, Success>>,
+        IRequestHandler<DeletedQuery, DeletedViewModel>
+    {
+        private readonly ICosmosDbQueryDispatcher _cosmosDbQueryDispatcher;
+        private readonly ISqlQueryDispatcher _sqlQueryDispatcher;
+        private readonly IProviderOwnershipCache _providerOwnershipCache;
+        private readonly JourneyInstanceProvider _journeyInstanceProvider;
+
+        public Handler(
+            ICosmosDbQueryDispatcher cosmosDbQueryDispatcher,
+            ISqlQueryDispatcher sqlQueryDispatcher,
+            IProviderOwnershipCache providerOwnershipCache,
+            JourneyInstanceProvider journeyInstanceProvider)
+        {
+            _cosmosDbQueryDispatcher = cosmosDbQueryDispatcher;
+            _sqlQueryDispatcher = sqlQueryDispatcher;
+            _providerOwnershipCache = providerOwnershipCache;
+            _journeyInstanceProvider = journeyInstanceProvider;
+        }
+
+        public async Task<OneOf<NotFound, ViewModel>> Handle(Query request, CancellationToken cancellationToken)
+        {
+            var venue = await _cosmosDbQueryDispatcher.ExecuteQuery(new GetVenueById { VenueId = request.VenueId });
+
+            if (venue == null)
+            {
+                return new NotFound();
+            }
+
+            var getCourses = _cosmosDbQueryDispatcher.ExecuteQuery(new GetAllCoursesForProvider { ProviderUkprn = venue.Ukprn, CourseStatuses = CourseStatus.Live | CourseStatus.Pending | CourseStatus.BulkUploadPending | CourseStatus.BulkUploadReadyToGoLive | CourseStatus.APIPending | CourseStatus.APIReadyToGoLive | CourseStatus.MigrationPending | CourseStatus.MigrationReadyToGoLive });
+            var getApprenticeships = _cosmosDbQueryDispatcher.ExecuteQuery(new GetApprenticeships { Predicate = a => a.ProviderUKPRN == venue.Ukprn && a.ApprenticeshipLocations.Any(al => al.VenueId == request.VenueId) });
+            var getTLevels = _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelsForProvider { ProviderId = request.ProviderId });
+
+            await Task.WhenAll(getCourses, getApprenticeships, getTLevels);
+
+            return CreateViewModel(venue, request.ProviderId, await getCourses, (await getApprenticeships).Values, await getTLevels);
```

Comment:
I don't understand why these are awaited twice.
I don't see an obvious cache of the result here. Won't it just run it all twice and double the number of requests?

Reply:
If you await an already completed `Task`, you get the result. The task is initially awaited by `await.Task.WhenAll(...)` - when that's completed, the tasks are completed and can be awaited again to yield their result.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/

Reply:
today I learned...
thanks